internal/sourcemap/sourcemap_fetcher.go

```diff
@@ -0,0 +1,125 @@
+// Licensed to Elasticsearch B.V. under one or more contributor
+// license agreements. See the NOTICE file distributed with
+// this work for additional information regarding copyright
+// ownership. Elasticsearch B.V. licenses this file to you under
+// the Apache License, Version 2.0 (the "License"); you may
+// not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+package sourcemap
+
+import (
+	"context"
+	"fmt"
+	"net/url"
+
+	"github.com/go-sourcemap/sourcemap"
+
+	"github.com/elastic/apm-server/internal/logs"
+	"github.com/elastic/elastic-agent-libs/logp"
+)
+
+type SourcemapFetcher struct {
+	metadata MetadataFetcher
+	backend  Fetcher
+	logger   *logp.Logger
+}
+
+func NewSourcemapFetcher(metadata MetadataFetcher, backend Fetcher) *SourcemapFetcher {
+	s := &SourcemapFetcher{
+		metadata: metadata,
+		backend:  backend,
+		logger:   logp.NewLogger(logs.Sourcemap),
+	}
+
+	return s
+}
+
+func (s *SourcemapFetcher) Fetch(ctx context.Context, name, version, path string) (*sourcemap.Consumer, error) {
+	original := Identifier{name: name, version: version, path: path}
+
+	select {
+	case <-s.metadata.Ready():
+		// the mutex is shared by the update goroutine, we need to release it
+		// as soon as possible to avoid blocking updates.
+		if i, ok := s.metadata.GetID(original); ok {
+			// Only fetch from ES if the sourcemap id exists
+			return s.fetch(ctx, i)
+		}
+	default:
+		s.logger.Debugf("Metadata cache not populated. Falling back to backend fetcher for id: %s, %s, %s", name, version, path)
+		// init is in progress, ignore the metadata cache and fetch the sourcemap directly
+		// return if we get a valid sourcemap or an error
+		if c, err := s.backend.Fetch(ctx, original.name, original.version, original.path); c != nil || err != nil {
+			return c, err
+		}
+
```

Comment:
I am missing why the following part is necessary. If the metadata fetcher is not done, `s.backend.Fetch` is triggered. If `c == nil && err == nil` it means that the sourcemap doesn't exist. What is the reason for blocking and waiting until metadata is ready etc?  

Especially if ES is just not reachable, this might lead to blocking any request.

Reply:
> I am missing why the following part is necessary. If the metadata fetcher is not done, s.backend.Fetch is triggered. If c == nil && err == nil it means that the sourcemap doesn't exist. What is the reason for blocking and waiting until metadata is ready etc?

If `c== nil && err == nil` it means there's no sourcemap associated with the provided name/version/path but the sourcemap could be an alias, in which case we block to populate the aliases: https://github.com/elastic/apm-server/pull/9722/files#diff-2941f7040f3ffdd6900280715c51cf111d7e6c28b0e7ba8449fdcd185a1dcaadR68
We're also checking the request context to avoid blocking requests indefinitely: https://github.com/elastic/apm-server/pull/9722/files#diff-2941f7040f3ffdd6900280715c51cf111d7e6c28b0e7ba8449fdcd185a1dcaadR71